plugins/ulimit-adjuster/adjuster.go

```diff
@@ -0,0 +1,161 @@
+		"NICE":       {},
+		"NOFILE":     {},
+		"NPROC":      {},
+		"RSS":        {},
+		"RTPRIO":     {},
+		"RTTIME":     {},
+		"SIGPENDING": {},
+		"STACK":      {},
+	}
+)
+
+func main() {
+	var (
+		pluginName string
+		pluginIdx  string
+		verbose    bool
+		opts       []stub.Option
+	)
+
+	l := logrus.StandardLogger()
+	l.SetFormatter(&logrus.TextFormatter{
+		PadLevelText: true,
+	})
+
+	flag.StringVar(&pluginName, "name", "", "plugin name to register to NRI")
+	flag.StringVar(&pluginIdx, "idx", "", "plugin index to register to NRI")
+	flag.BoolVar(&verbose, "verbose", false, "enable (more) verbose logging")
+	flag.Parse()
+	ctx := log.WithLogger(context.Background(), l.WithField("name", pluginName).WithField("idx", pluginIdx))
+	log.G(ctx).WithField("verbose", verbose).Info("starting plugin")
+
+	if verbose {
+		l.SetLevel(logrus.DebugLevel)
+	}
+
+	if pluginName != "" {
+		opts = append(opts, stub.WithPluginName(pluginName))
+	}
+	if pluginIdx != "" {
+		opts = append(opts, stub.WithPluginIdx(pluginIdx))
+	}
+
+	p := &plugin{l: log.G(ctx)}
+	var err error
+	if p.stub, err = stub.New(p, opts...); err != nil {
+		log.G(ctx).Fatalf("failed to create plugin stub: %v", err)
+	}
+
+	if err := p.stub.Run(context.Background()); err != nil {
+		log.G(ctx).Errorf("plugin exited with error %v", err)
+		os.Exit(1)
+	}
+}
+
+type plugin struct {
+	stub stub.Stub
+	l    *logrus.Entry
+}
+
+func (p *plugin) CreateContainer(
+	ctx context.Context,
+	pod *api.PodSandbox,
+	container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+	if pod != nil {
+		p.l = p.l.WithField("pod", pod.Name)
+	}
+	if container != nil {
+		p.l = p.l.WithField("container", container.Name)
+	}
+	ctx = log.WithLogger(ctx, p.l)
+	log.G(ctx).Debug("create container")
+
+	ulimits, err := parseUlimits(ctx, container.Name, pod.Annotations)
+	if err != nil {
+		log.G(ctx).WithError(err).Debug("failed to parse annotations")
+		return nil, nil, err
+	}
+	adjust := &api.ContainerAdjustment{}
+	for _, u := range ulimits {
+		log.G(ctx).WithField("type", u.Type).WithField("hard", u.Hard).WithField("soft", u.Soft).Debug("adjust rlimit")
+		adjust.AddRlimit(u.Type, u.Hard, u.Soft)
+	}
+	return adjust, nil, nil
+}
+
+type ulimit struct {
+	Type string `json:"type"`
+	Hard uint64 `json:"hard"`
+	Soft uint64 `json:"soft"`
+}
+
+func parseUlimits(ctx context.Context, container string, annotations map[string]string) ([]ulimit, error) {
+	key := ulimitKey + "/container." + container
+	val, ok := annotations[key]
+	if !ok {
+		log.G(ctx).Debugf("no annotations found with key %q", key)
+		return nil, nil
+	}
+	ulimits := make([]ulimit, 0)
+	if err := yaml.Unmarshal([]byte(val), &ulimits); err != nil {
+		return nil, err
+	}
+	for i := range ulimits {
+		u := ulimits[i]
+		typ := strings.TrimPrefix(strings.ToUpper(u.Type), rlimitPrefix)
+		if _, ok := valid[typ]; !ok {
+			log.G(ctx).WithField("raw", u.Type).WithField("trimmed", typ).Debug("failed to parse type")
```

Comment:
maybe output `typ` to debug output

Reply:
`WithField("raw", u.Type).WithField("trimmed", typ)` that outputs `typ`.